File: src/main/scala/lang/Scanner.scala
package cql.lang

import scala.util.Try

class Scanner(program: String):
  var tokens: List[Token] = List.empty
  var start = 0
  var current = 0
  var line = 1
  var hasError = false;
  var reservedChars = "+@:()\""

  def scanTokens: List[Token] =
    while (!isAtEnd) {
      // We are at the beginning of the next lexeme.
      start = current;
      scanToken
    }

    println(tokens.map(_.lexeme))
    println(joinStringTokens(tokens).map(_.lexeme))

    joinStringTokens(tokens) :+ Token(TokenType.EOF, "", None, current, current)

  def joinStringTokens(tokens: List[Token]) = tokens.foldLeft(
    List.empty[Token]
  )((acc, token) => {
    (acc.lastOption, token.tokenType) match {
      case (Some(prevToken), TokenType.STRING)
          if prevToken.tokenType == TokenType.STRING =>
        acc.slice(0, acc.length - 1) :+ Token(
          TokenType.STRING,
          lexeme = prevToken.lexeme + " " + token.lexeme,
          literal = Some(
            prevToken.literal.getOrElse("") + " " + token.literal.getOrElse("")
          ),
          prevToken.start,
          token.end
        )
      case _ => acc :+ token
    }
  })

  def isAtEnd = current == program.size

  def scanToken =
    advance match {
      case '+' =>
        addKey(TokenType.QUERY_FIELD_KEY)
      case '@' =>
        addKey(TokenType.QUERY_OUTPUT_MODIFIER_KEY)
      case ':' =>
        addValue
      case '(' =>
        addToken(TokenType.LEFT_BRACKET)
      case ')' =>
        addToken(TokenType.RIGHT_BRACKET)
      case ' '  => ()
      case '\r' => ()
      case '\t' => ()
      case '\n' =>
        line = line + 1
      case '"'                 => addString
      case _ if isReservedWord => addIdentifier
      case _                   => addUnquotedString
    }

  def addKey(tokenType: TokenType) =
    while ((peek != ':' && !peek.isWhitespace) && !isAtEnd)
      advance

    if (current - start == 1) addToken(tokenType)
    else
      val key = program.substring(start + 1, current)
      addToken(tokenType, Some(key))

  def addValue =
    while ((!peek.isWhitespace) && !isAtEnd)
      advance

    if (current - start == 1) addToken(TokenType.QUERY_VALUE, None)
    else
      val value = program.substring(start + 1, current)
      addToken(TokenType.QUERY_VALUE, Some(value))

  def isReservedWord =
    Token.reservedWords.exists { case (str, _) =>
      program.substring(start).startsWith(str)
    }

  def addIdentifier =
    while (peek.isLetterOrDigit) advance
    val text = program.substring(start, current)
    Token.reservedWords.get(text) match {
      case Some(token) => addToken(token)
      case None        => error(line, "Expected identifier")
    }

  def addUnquotedString =
    while (!peek.isWhitespace && peek != ')' && !isAtEnd)
      advance

    addToken(TokenType.STRING, Some(program.substring(start, current)))

  def addString =
    while ((peek != '"') && !isAtEnd)
      advance

    if (isAtEnd) error(line, "Unterminated string at end of file")
    else
      advance
      addToken(
        TokenType.STRING,
        Some(program.substring(start + 1, current - 1))
      )

  def addToken(tokenType: TokenType, literal: Option[String] = None) =
    val text = program.substring(start, current)
    println(s"text${text}")
    tokens = tokens :+ Token(tokenType, text, literal, start, current - 1)

  def advance =
    val previous = current
    current = current + 1
    program(previous)

  def matchChar(expected: Char) =
    if (isAtEnd || program(current) != expected) false
    else
      current = current + 1
      true

  def peek = if (isAtEnd) '\u0000' else program(current)
  def peekNext =
    if (current + 1 >= program.size) '\u0000' else program.charAt(current + 1)

  def error(line: Int, message: String) = report(line, "", message)

  def report(line: Int, where: String, message: String) =
    println(s"[line ${line}] Error${where}: ${message}")
    hasError = true;

File: prosemirror-client/src/lang/scanner.ts
import { Token, TokenType } from "./token";
import { isLetterOrDigit, isWhitespace } from "./util";

export class Scanner {
  private tokens: Array<Token> = [];
  private start = 0;
  private current = 0;
  private line = 1;

  constructor(private program: string) {}

  public scanTokens = (): Token[] => {
    while (!this.isAtEnd()) {
      // We are at the beginning of the next lexeme.
      this.start = this.current;
      this.scanToken();
    }

    return this.tokens.concat(
      new Token(TokenType.EOF, "", undefined, this.current, this.current)
    );
  };

  private scanToken = () => {
    switch (this.advance()) {
      case "+":
        this.addKey(TokenType.CHIP_KEY);
        return;
      case ":":
        this.addValue();
        return;
      case "(":
        this.addToken(TokenType.LEFT_BRACKET);
        return;
      case ")":
        this.addToken(TokenType.RIGHT_BRACKET);
        return;
      case " ":
        return;
      case "\r":
      case "\t":
      case '"':
        this.addString();
        return;
      default:
        this.addIdentifierOrUnquotedString();
        return;
    }
  };

  private addKey = (tokenType: TokenType) => {
    while (this.peek() != ":" && !isWhitespace(this.peek()) && !this.isAtEnd())
      this.advance();

    if (this.current - this.start === 1) this.addToken(tokenType);
    else {
      const key = this.program.substring(this.start + 1, this.current);

      this.addToken(tokenType, key);
    }
  };

  private addValue = () => {
    while (!isWhitespace(this.peek()) && !this.isAtEnd()) this.advance();

    if (this.current - this.start == 1) {
      this.addToken(TokenType.CHIP_VALUE);
    } else {
      const value = this.program.substring(this.start + 1, this.current);
      this.addToken(TokenType.CHIP_VALUE, value);
    }
  };

  private addIdentifierOrUnquotedString = () => {
    while (isLetterOrDigit(this.peek())) {
      this.advance();
    }

    const text = this.program.substring(this.start, this.current);
    const maybeReservedWord =
      Token.reservedWordMap[text as keyof typeof Token.reservedWordMap];

    return maybeReservedWord
      ? this.addToken(maybeReservedWord)
      : this.addUnquotedString();
  };

  private addUnquotedString = () => {
    while (
      // Consume whitespace up until the last whitespace char
      (!isWhitespace(this.peek()) ||
        isWhitespace(this.peek(1)) ||
        this.isAtEnd(1)) &&
      this.peek() != ")" &&
      !this.isAtEnd()
    ) {
      this.advance();
    }

    this.addToken(
      TokenType.STRING,
      this.program.substring(this.start, this.current)
    );
  };

  private addString = () => {
    while (this.peek() != '"' && !this.isAtEnd()) {
      this.advance();
    }

    if (this.isAtEnd()) {
      this.error(this.line, "Unterminated string at end of file");
    } else {
      this.advance();
    }

    this.addToken(
      TokenType.STRING,
      this.program.substring(this.start + 1, this.current - 1)
    );
  };

  private addToken = (tokenType: TokenType, literal?: string) => {
    const text = this.program.substring(this.start, this.current);
    this.tokens = this.tokens.concat(
      new Token(tokenType, text, literal, this.start, this.current - 1)
    );
  };

  private advance = () => {
    const previous = this.current;
    this.current = this.current + 1;
    return this.program[previous];
  };

  private peek = (offset: number = 0) =>
    this.program[this.current + offset] === undefined
      ? "\u0000"
      : this.program[this.current + offset];

  private isAtEnd = (offset: number = 0) =>
    this.current + offset === this.program.length;

  private error = (line: number, message: String) =>
    this.report(line, "", message);

  private report = (line: number, where: String, message: String) => {
    console.log(`[line ${line}] Error${where}: ${message}`);
  };
}
